### DISCUSSION TITLE
为什么 NUMA 影响程序的延迟 - 面向信仰编程 · /whys-the-design-numa-performance

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#630] [Announcements]
https://draveness.me/whys-the-design-numa-performance/

**Comment by leafduo** [comment]
感觉 Kubernetes 这一层应该解决一部分问题，不过好像 Kubernetes 的 NUMA 支持还在比较早期的阶段

**Comment by draveness** [comment]
> @leafduo 感觉 Kubernetes 这一层应该解决一部分问题，不过好像 Kubernetes 的 NUMA 支持还在比较早期的阶段

是的，Kuberentes 不支持 NUMA 感知调度，只是在 Kubelet 层面支持几种 NUMA 策略

**Comment by leafduo** [comment]
> @draveness
> > @leafduo 感觉 Kubernetes 这一层应该解决一部分问题，不过好像 Kubernetes 的 NUMA 支持还在比较早期的阶段
> 
> 是的，Kuberentes 不支持 NUMA 感知调度，只是在 Kubelet 层面支持几种 NUMA 策略

意思是只有在一个 node 的范围中调度，不能在集群的范围中调度？

**Comment by draveness** [comment]
> @leafduo
> > @draveness
> > > @leafduo 感觉 Kubernetes 这一层应该解决一部分问题，不过好像 Kubernetes 的 NUMA 支持还在比较早期的阶段
> > 
> > 是的，Kuberentes 不支持 NUMA 感知调度，只是在 Kubelet 层面支持几种 NUMA 策略
> 
> 意思是只有在一个 node 的范围中调度，不能在集群的范围中调度？

是的，只有 kubelet 才能决定是不是运行在 NUMA 上，但是如果 node 不支持或者无法把 Pod 分配到 NUMA 上， kubelet 会在 pod admission phase 报错或者运行在不同 NUMA 节点上

**Comment by et-zhou** [comment]
“能够页面频繁换入换出的可能性”这句感觉不太通顺，是不是少了一个“降低“？

---

*2020-10-26 UPDATES: 已修复*

```diff --git a/insight/content/post/2020-10-23-whys-the-design-numa-performance.md b/insight/content/post/2020-10-23-whys-the-design-numa-performance.md
index 2fc82c6..ccd34d4 100644
--- a/insight/content/post/2020-10-23-whys-the-design-numa-performance.md
+++ b/insight/content/post/2020-10-23-whys-the-design-numa-performance.md
@@ -103,7 +103,7 @@

-想要解决该问题，我们需要使用上一节提到的 `numactl` 将内存的分配策略改为 `interleave`，使用该内存分配策略会使得 MySQL 的内存均匀地分配到不同的 NUMA 节点上，能够页面频繁换入换出的可能性。
+想要解决该问题，我们需要使用上一节提到的 `numactl` 将内存的分配策略改为 `interleave`，使用该内存分配策略会使得 MySQL 的内存均匀地分配到不同的 NUMA 节点上，能够降低页面频繁换入换出的可能性。

 ![even-memory-node](https://img.draveness.me/2020-10-24-16035525564395/even-memory-node.png)
```

**Comment by npuichigo** [comment]
"将缓存池中的数据按照块或者索引只能地分配到不同节点上；"是笔误吗

---

*2021-03-01 UPDATES: 已修复*

**Comment by plantegg** [comment]
好文章， 文中提到的 Jeremy Cole 的文章应该搞错了一些事情：https://plantegg.github.io/2021/05/14/%E5%8D%81%E5%B9%B4%E5%90%8E%E6%95%B0%E6%8D%AE%E5%BA%93%E8%BF%98%E6%98%AF%E4%B8%8D%E6%95%A2%E6%8B%A5%E6%8A%B1NUMA/

**Comment by dirtysalt** [comment]
> @plantegg
> 好文章， 文中提到的 Jeremy Cole 的文章应该搞错了一些事情：https://plantegg.github.io/2021/05/14/%E5%8D%81%E5%B9%B4%E5%90%8E%E6%95%B0%E6%8D%AE%E5%BA%93%E8%BF%98%E6%98%AF%E4%B8%8D%E6%95%A2%E6%8B%A5%E6%8A%B1NUMA/

好文章，所以现在linux在numa上默认配置大部分时候都是够用的了，特定情况下需要绑定或者是自己做numa分配。

**Reply by plantegg** [reply]
总结就是：
- 任何时候都要在BIOS中打开NUMA，把结构信息透传给OS；
- 默认参数下就近分配内存性能最好；
- 如果能做绑核能避免调度到其它numa node后跨node访问原来的内存，另外也提升了cache命中率，绑核有奇效

**Comment by SamZhangQingChuan** [comment]
应该是多处理器而不是多核？

**Reply by draveness** [reply]
可以详细展开一下么

**Reply by SamZhangQingChuan** [reply]
如果只有一个处理器，没有多个总线来做成numa架构？

**Reply by plantegg** [reply]
和多处理器没有关系，和核的数量多少有关系，也就是核数多了没法一个总线连所有内存，实际现在的服务器内存都是直连CPU不再走总线（CPU芯片上直接做好了PCI接口）。多看看基本概念，比如这篇 [https://plantegg.github.io/2021/06/01/CPU%E7%9A%84%E5%88%B6%E9%80%A0%E5%92%8C%E6%A6%82%E5%BF%B5/](https://plantegg.github.io/2021/06/01/CPU%E7%9A%84%E5%88%B6%E9%80%A0%E5%92%8C%E6%A6%82%E5%BF%B5/)

**Reply by SamZhangQingChuan** [reply]
好的谢谢大佬
